Storefront/CartViewController.swift

```diff
@@ -111,12 +111,52 @@ class CartViewController: ParallaxViewController {
     // ----------------------------------
     //  MARK: - Actions -
     //
-    func openSafariFor(_ checkout: CheckoutViewModel) {
+    func openWKWebViewControllerFor(_ checkout: CheckoutViewModel) {
         let webController = WebViewController(url: checkout.webURL, accessToken: AccountController.shared.accessToken)
         webController.navigationItem.title = "Checkout"
         self.navigationController?.pushViewController(webController, animated: true)
     }
     
+    func openSFSafariViewControllerForURL(_ url: URL) {
+        self.navigationController?.isNavigationBarHidden = true
+        let webController = SFSafariViewController(url: url)
+        self.navigationController?.pushViewController(webController, animated: true)
+    }
+    
+    func buildShopPayURL(_ shopURL: URL, cartItems: [CartItem]) -> URL? {
+        func decodeBase64String(_ base64String: String) -> String {
+            let decodedData = Data(base64Encoded: base64String)!
+            return String(data: decodedData, encoding: .utf8)!
+        }
+        func extractVariantId(_ fullVariantId: String) -> String {
+            // Example string: gid://shopify/ProductVariant/31384149360662
+            let pattern = #"gid://shopify/ProductVariant/(\d+)"#
+            let regex = try! NSRegularExpression(pattern: pattern, options: [])
+            let result = regex.matches(in:fullVariantId, range:NSMakeRange(0, fullVariantId.utf16.count))
+            if (result.isEmpty) {
+                // Handle error cases.
+                return ""
+            }
+            if let substringRange = Range(result[0].range(at: 1), in: fullVariantId) {
+                    return String(fullVariantId[substringRange])
+                }
+            return ""
+        }
+        func buildVariantSlugFor(_ item: CartItem) -> String {
+            return extractVariantId(decodeBase64String(item.variant.id)) + ":" + String(item.quantity)
+        }
+        
+        // Build a Shop Pay checkout link.
+        var components = URLComponents()
+        components.scheme = "https"
+        components.host = shopURL.host
+        components.path = "/cart/" + cartItems.map{buildVariantSlugFor($0)}.joined(separator: ",")
```

Comment:
We should be able to pass the `buildVariantSlugFor` function directly to `map` here:

```suggestion
        components.path = "/cart/" + cartItems.map(buildVariantSlugFor).joined(separator: ",")
```